### ISSUE TITLE
Right side border missing in pdf (for foreign language)

### ISSUE THREAD

**Issue by sunithapremkumar** [danfickle/openhtmltopdf#524]
I am generating pdf from html. The html file looks good but in my pdf I see the right side border is missing. 
Please help. Attached the partial image of the same

<img width="303" alt="Screen Shot 2020-08-04 at 7 10 19 AM" src="https://user-images.githubusercontent.com/22286174/89304028-cffdc100-d621-11ea-969a-f327c0293d00.png">

**Comment by danfickle** [comment]
Possibly try `box-sizing: border-box;` on the Element which has a missing right border.

**Comment by sunithapremkumar** [comment]
Thanks for the quick response.  The html is generated by some other tool where we wouldn't have control on the tags. Is there a way to do some configuration during pdf generation which takes care of this boxing?  Below is my code snipper

private byte[] convertHtmlToPdf(byte[] htmlFile) throws Exception {
        //clean the html if the html contains any non closing tags as the conversion library expects it to be in proper format
        String cleaned_html = cleanHtml(htmlFile);

        ByteArrayOutputStream os = new ByteArrayOutputStream();
        PdfRendererBuilder builder = new PdfRendererBuilder();
        builder.useFastMode();
        builder.withHtmlContent(cleaned_html, ".");
        builder.toStream(os);
        builder.run();

        return os.toByteArray();
    }

**Comment by syjer** [comment]
hi @sunithapremkumar 

> The html is generated by some other tool where we wouldn't have control on the tags.

you can parse the html and add in the DOM tree a custom `<style>` element with the missing/correct css...

But maybe, it's even better if you extract the data from the html and provide a custom html that _you_ have control filled with the extracted data.

**Comment by sunithapremkumar** [comment]
Looks like adding boxing as below is still not helping

` <tr height="17" style="height: 12.75pt;box-sizing: border-box" id="r41">
`
Attached the html and pdf output. I clearly see the right border style is 'solid' but still not taking effect. Suspecting if this is due to non- english characters in pdf (as this is german locale)

[Archive.zip](https://github.com/danfickle/openhtmltopdf/files/5022844/Archive.zip)

**Comment by syjer** [comment]
hi @sunithapremkumar , I had a look at the html you provided, the main issue is the fact the tables/col&co all have a specified width in pt.

For example: 

 - the main table has `width="674" style="border-collapse: collapse; table-layout: fixed; width: 505pt;`
 - there are various `<col class="x16" width="237" style="background-color: initial; background-image: none; width: 177.75pt;">`

having a fixed width will clash with the page size and will go outside.

As an experiment, I removed all the width (both in the style attribute and as a attribute) and set a width of 100% for the table and I got something like that as an output:

![pdfdoc](https://user-images.githubusercontent.com/498146/89422597-d54c3000-d735-11ea-867a-296aee89c83a.png)

So basically, I would advise to not provide the raw html to openhtmltopdf, but instead to create a new one that has a correct css and fill your data inside.

**Comment by sunithapremkumar** [comment]
You are right after removing the main table width it works. But I have other html, same like this  but in different locale, it does work with fixed width table.  Attached the same. Trying to understand how it works in the attached html but not in the previous one (although both have fixed width).

[UK.zip](https://github.com/danfickle/openhtmltopdf/files/5031376/UK.zip)

**Comment by syjer** [comment]
FYI the attached zip contains the same German pdf/html as your previous messages.

But anyway, most likely, as it seems generated from an excel file, the fixed size (attributes and css rule) on the one that worked were just right to fit the width of the page (for example, due to shorter words vs the German one).

**Comment by sunithapremkumar** [comment]
I am sorry, attaching the right files.   Initially I even thought that same way as you that the data in UK might be shorter to fit in the size. But the files doesn't have any data it has only the headers now, I don't think the headers of German are any longer than UK

[UK.zip](https://github.com/danfickle/openhtmltopdf/files/5035223/UK.zip)

By the way, if I have to provide multilingual support in my report say if I have to display the data in other languages, can you please tell me what do I need to do? I believe this library supports multilingual

**Comment by syjer** [comment]
> By the way, if I have to provide multilingual support in my report say if I have to display the data in other languages, can you please tell me what do I need to do? I believe this library supports multilingual

Except ensuring you load the required fonts and in case of RTL language include the correct support (see https://github.com/danfickle/openhtmltopdf/wiki/Integration-Guide#support-for-bi-directional-rtl-and-shaped-text ), you don't need to do anything special.

In the end, this library only generate a pdf of your html.

**Comment by syjer** [comment]
by the way, the table in the uk html has specified:

```html
<table border="0" cellpadding="0" cellspacing="0" width="624" style="border-collapse: collapse; table-layout: fixed; width: 468pt;">
```
 
468pt may be small enough to fit the page, compared to the 505pt of the German one.

**Comment by sunithapremkumar** [comment]
Regarding multilingual support. I tried the below code as suggested in the link. It is still not working for russian language. Attached the html containing russian language.

[report-russia.zip](https://github.com/danfickle/openhtmltopdf/files/5058639/report-russia.zip)


```
   private static void convertHtmlToPDfusingOpenHtmlToPdf(String html) throws Exception{
try (OutputStream os = new FileOutputStream("/output/main-report.pdf")) {
            PdfRendererBuilder builder = new PdfRendererBuilder();
            builder.useUnicodeBidiSplitter(new ICUBidiSplitter.ICUBidiSplitterFactory());
            builder.useUnicodeBidiReorderer(new ICUBidiReorderer());
            builder.defaultTextDirection(BaseRendererBuilder.TextDirection.LTR); // OR RTL
            builder.useFastMode();
            builder.withHtmlContent(html, ".");
            builder.toStream(os);
            builder.run();
        }
}
```

**Comment by syjer** [comment]
there is only the html in your zip.

Could please explain what does not work exactly ?

edit: note: you will most likely need to import a Cyrillic font.

**Comment by sunithapremkumar** [comment]
If you see the html it does shows up the russian characters properly.. but pdf output after conversion contains #### as opposed to actual russian characters in html

Can you please explain how to use that font? Is that the font specific for russian language? As we also deal with other languages

**Comment by syjer** [comment]
the '#' character mean you don't have a font that can handle the specific character.

See https://github.com/danfickle/openhtmltopdf/wiki/Fonts .

In your specific case, you will need to programmatically add the right font under the 'Arial' name.

See https://fonts.google.com/  for selecting a font for a given language. In your case I would advise, as the wiki page suggest, to use the Noto font family and load the correct `.ttf` for the specified language.

edit: Noto Sans should also support out of the box the Cyrillic alphabet. But you will need to load another `.ttf` for CJK, and other scripts too, like arabic, hebrew...

**Comment by sunithapremkumar** [comment]
Do you mean add those fonts in html as well as configure it in PdfRendererBuilder?  While adding in html do we define the font in style and refer that style everywhere it is  used?  If this is the case, its very hard to make the new style to be used all over in html..  Question is if html can display the characters properly why would I modify the html just for the pdf to display properly?

Moreover below is the font used in html for those russian text.. Won't your library picks it from this font and generates the pdf? 
x32 { vertical-align: bottom; white-space: nowrap; background-color: initial; background-image: initial; font-size: 10pt; font-weight: 400; font-style: normal; font-family: "Arial", "sans-serif"; border-top-color: windowtext; border-top-style: solid; border-top-width: 1px; border-right-color: windowtext; border-right-style: solid; border-right-width: 1px; border-bottom-color: windowtext; border-bottom-style: solid; border-bottom-width: 1px; border-left-color: windowtext; border-left-style: solid; border-left-width: 1px; }

Can you please give me the exact font solution just for this russian html to display correctly? what changes to be made in code and html? I will try it out.. but changing entire html would not be feasible solution

**Comment by syjer** [comment]
hi @sunithapremkumar ,

It's clearly documented in the wiki page what you need to do when you have the '#' symbols. You don't need to modify the html/css. You only need in the **java** code to import the correct **TTF** font with the "Arial" name, nothing more, nothing less.

Something like that: 
```java
builder.useFont(new File("/path-to-fonts/fonts/MyFont.ttf"), "Arial");
```


As to why you need to import the font: the PDF standard out of the box include only some fonts and they cover a small range of characters. That's why for most non western language you will need to import an external font. Additionally, this library will not try to find the system fonts (as you may imagine, what fonts you have on your machine may vary to what is available when you deploy on a server). 

Please note that I can't do more than that, or else I would be doing your job.

Frankly, I would advise to first understand the library before trying to feed it a complex html/css.

**Comment by sunithapremkumar** [comment]
Hi @syjer 
     I am kind of new to this Fonts, so was asking basic questions.. I have gone through bit about fonts and added 'Noto Sans' Font which I see supports most of the languages. Below is the font setting I configured with which the report shows up the russian characters which am glad to see. 

     
  ```
builder.useFont( new File(Test.class.getResource("/fonts/NotoSans-Bold.ttf").getFile()), "Arial");
builder.useFont( new File(Test.class.getResource("/fonts/NotoSans-BoldItalic.ttf").getFile()), "Arial");
builder.useFont( new File(Test.class.getResource("/fonts/NotoSans-Regular.ttf").getFile()), "Arial");
builder.useFont( new File(Test.class.getResource("/fonts/NotoSans-Italic.ttf").getFile()), "Arial");
```

Note - I also loaded NotoSansCJK fonts to render chinese character but it doesn't display which you can notice as ### in pdf in first row.  

Just a couple of minor issues in the russian pdf report , I do see the size of label goes over the column border (if you notice the 7th row within the first table in pdf). Also the very first element right under Global Assert Protection, there is a data  (john) but it's not shown as the label is written on top of it.  You can see these 2 issues doesn't exist in html but these are incorrectly rendered in pdf.. 
I remember we talked about the fixed width earlier after which I did fix that fixed width issue now it's no longer fixed width but still these issues. Attached the html and pdf output for your reference.
[Russia-report1.zip](https://github.com/danfickle/openhtmltopdf/files/5061482/Russia-report1.zip)

One more question on font loading api. if I want to load as InputStream instead of File, I checked that  buidler.usefont()  takes in FSSupplier via which we can pass in stream. but couldn't the implementation of it to use. Is it possible to supply InoutStream instead of File?

thanks for your help as we have production release coming up we are trying to fix these issues.. your help is greatly appreciated.

**Comment by syjer** [comment]
if you load multiple variants of the same fonts under the same name (like you are doing) you need to use the more complete method, or else you are overriding the definition, see https://javadoc.io/doc/com.openhtmltopdf/openhtmltopdf-core/latest/com/openhtmltopdf/outputdevice/helper/BaseRendererBuilder.html#useFont(java.io.File,java.lang.String,java.lang.Integer,com.openhtmltopdf.outputdevice.helper.BaseRendererBuilder.FontStyle,boolean)

Anyway, in your specific case you can only import the Regular variant. I think it's simpler.

>Note - I also loaded NotoSansCJK fonts to render chinese character but it doesn't display which you can notice as ### in pdf in first row.

This one is more complex, you will most likely need to modify the css/html.

> Is it possible to supply InoutStream instead of File?

yes you can. You just implement the following interface https://javadoc.io/doc/com.openhtmltopdf/openhtmltopdf-core/latest/com/openhtmltopdf/extend/FSSupplier.html . 

For the difference in output, you should not expect 1:1 result from your browser of choice. This is expected. You will need to tweak your html/css.

**Comment by sunithapremkumar** [comment]
Thanks for your quick response. I tried implementing FSSupplier, it works as expected.

By the way for the data over lapping issue which I mentioned in earlier message (Stated below as well) do you know why its happening??

```
Just a couple of issues in the russian pdf report , I do see the size of label goes over the column border (if you notice the 7th row within the first table in pdf). Also the very first element right under Global Assert Protection, there is a data (john) but it's not shown as the label is written on top of it. You can see these 2 issues doesn't exist in html but these are incorrectly rendered in pdf..
I remember we talked about the fixed width earlier after which I did fix that fixed width issue now it's no longer fixed width but still these issues. Attached the html and pdf output for your reference.
```
Attaching screenshot of the same pdf which I sent in earlier message.
<img width="891" alt="Russian-lang" src="https://user-images.githubusercontent.com/22286174/90057111-c95ff100-dc94-11ea-88c9-21616b49882b.png">


For chinese characters are you saying that we cannot fix in java to render, do we need to update html?  any reason why is this treated differently than russian? As long as we have the fonts supporting Chinese characters loaded in builder, will it not work?

General question, before even loading those Noto font for russia, this library works well for spanish/italy/french languages. Only for russia I have loaded the font. So wondering for those other languages, does the library have the ttf bundled with it? How does it work automatically for other languages?

**Comment by syjer** [comment]
>By the way for the data over lapping issue which I mentioned in earlier message (Stated below as well) do you know why its happening??

most likely because you have fixed width for the column in your table. That's why , as written various time before, it would be better if you write a custom html.

 >For chinese characters are you saying that we cannot fix in java to render, do we need to update html? any reason why is this treated differently than russian? As long as we have the fonts supporting Chinese characters loaded in builder, will it not work?

You can try, but as far as I know, you will need to do something like https://sandbox.openhtmltopdf.com/?file=cjk.htm .

You need to load the CJK font under a different name than arial in your java code and add it everywhere where 'Arial' is also present in the css.

> So wondering for those other languages, does the library have the ttf bundled with it? 

No, in the pdf standard by default you have Helvetica/Times/Courrier/Symbol available that you can use without providing a custom font, unfortunately they cover a rather narrow character set (which english/french/italian/spanish use), so in most cases if you go outside of the "wester world" you will need to load the correct font. For example, Polish alphabet is not covered by default. 

Generally it's best practice to always use a custom font and not rely on the default one.

I'll repeat myself once more, it would be better if you make your own clean html from scratch, you will avoid a lot of problems...

**Comment by danfickle** [comment]
I think you could also modify the dom to inject a style tag as the last child of head that uses an !important font-family declaration to override the existing value.

However, in general expecting this library to work with unmodified pre existing HTML made for the browser is a long shot.

**Comment by sunithapremkumar** [comment]
I checked the html for data overlapping issue.. there is no fixed width for those rows and also in table but still it overlaps.. 

By the way this html is generated from excel using aspose library.  It's a bit long story. We have templates in excel which we are trying to standardize to html for which we Aspose.. Then we use freemarker to generate the reports out of this html template.  So the html is generated from excel currently using aspose.. I will have limited control to inject something static on this html during excel to html conversion but not too much on element level.   

The recommendation you are giving is for chinese issue? You mean add the CJK style at the end of the style declaration in header? I first tried adding this cjk style alone, it didn't work and then added x109 using this font family. Changed that particular tag (containing chinese) to use this x109 font.. still didn't work

```
* {
  font-family: 'cjk', 'deja-sans', sans-serif;
}
.x109 { text-align: left; vertical-align: middle; white-space: normal; word-wrap: break-word; background-color: initial; background-image: initial; font-size: 10pt; font-weight: 400; font-style: normal; font-family: "cjk", "sans-serif"; border-top-color: windowtext; border-top-style: solid; border-top-width: 1px; border-right-color: initial; border-right-style: none; border-right-width: initial; border-bottom-color: windowtext; border-bottom-style: solid; border-bottom-width: 1px; border-left-color: windowtext; border-left-style: solid; border-left-width: 1px; }
```
If possible you may give it a quick try suing the html I sent?   I prefer configuring builder to fix this if not I can inject add static style in html under styles tag but wouldn't be able to touch the specific html element containing the content.

We have production release in couple of weeks, trying to fix hese 2 issues overlapping and chinese.   We basically went for aspose for this conversion but they did have some issues and their turnaround time was too slow and so quickly moved over to this library

**Comment by syjer** [comment]
>We have production release in couple of weeks [...]

>We basically went for aspose for this conversion but they did have some issues and their turnaround time was too slow and so quickly moved over to this library

Sorry, but you should not try to impose your deadlines. Especially  as aspose is a paid library and you consider their turnaround too slow.

I think I've given enough hints and help in how to solve the various issues.

**Comment by sunithapremkumar** [comment]
Sorry my intention wasn't trying to impose the deadline to you at all in anyways..  If you see I have tried different ways you mentioned as well.  The provided solution is not working in both issues.. So was trying to seek help.

**Comment by danfickle** [comment]
OK, I downloaded the Russian version:
+ I had to add some closing tags. I assume you are using an HTML5 parser to do this.
+ Adding a `<style>` tag at the end of the body will not work, it has to be at end of head tag.
+ The overflow issue was because the table cells had `white-space: nowrap` set on them

Anyway, fixing the styles is as simple as:
````java
package com.example;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;

import com.openhtmltopdf.outputdevice.helper.BaseRendererBuilder.FontStyle;
import com.openhtmltopdf.pdfboxout.PdfRendererBuilder;

import org.w3c.dom.Element;
import org.w3c.dom.Node;

public final class App {
    private static final String EXTRA_STYLE = 
        "* { font-family: 'noto' !important; }\n" +
        "td { white-space: normal !important; }\n" +
        "body style { display: none !important; }\n";

    public static void main(String[] args) throws IOException {
        PdfRendererBuilder builder = new PdfRendererBuilder();
        
        try (FileOutputStream fos = new FileOutputStream("C:\\Users\\dan\\Desktop\\result.pdf")) {
            builder.withFile(new File("C:\\Users\\dan\\Desktop\\report.html"));
            builder.useFont(new File("C:\\Users\\dan\\Desktop\\fonts\\NotoSans-Regular.ttf"), "noto");
            builder.useFont(new File("C:\\Users\\dan\\Desktop\\fonts\\NotoSans-Bold.ttf"), "noto", 700, FontStyle.NORMAL, true);
            builder.toStream(fos);
            builder.useFastMode();

            builder.addDOMMutator(doc -> {
                Element style = doc.createElement("style");
                Node node = doc.createTextNode(EXTRA_STYLE);
                style.appendChild(node);
                doc.getElementsByTagName("head").item(0).appendChild(style);
            });

            builder.run();;
        }
    }
}
````

This produced the following PDF which looks correct, although I don't read Russian!

[result.pdf](https://github.com/danfickle/openhtmltopdf/files/5078606/result.pdf)


EDIT: For the Chinese to work, you'll have to add it in the extra style font list: `font-family: 'noto', 'cjk' !important;`. Also add it in the builder.

**Comment by sunithapremkumar** [comment]
@danfickle  Thank you very much for your response and trying it out

I knew that the no-wrap was one of the reason for overlapping column. But we do not want the first column to be wrapped as we are ok to wrap the second column (data column).  I was trying to find solution without wrapping first column instead adjusting the first column to fit to the size of the value inside it.. but so far no luck.

Regarding the chinese, I tried as you mentioned.. see below the code.. not sure what am missing.. still doesn't work

`private static final String EXTRA_STYLE =
            "* { font-family: 'noto', 'cjk' !important; }\n" +
                    "td { white-space: normal !important; }\n" +
                    "body style { display: none !important; }\n";
   
 private static void convertHtmlToPDfusingOpenHtmlToPdf(String html) throws Exception{
        try (OutputStream os = new FileOutputStream("/output/main-report.pdf")) {
            PdfRendererBuilder builder = new PdfRendererBuilder();
            builder.useFastMode();
            builder.withHtmlContent(html, ".");
            builder.toStream(os);
           fontList.stream().forEach(font -> {
                    builder.useFont(new FontSupplier("/fonts/NotoSans-Regular.ttf"), "noto");  //this is for showing russian chracters
            });
            builder.useFont(new FontSupplier("/fonts/NotoSansCJK-Regular.ttf"), "cjk"); //this is for showing chinese
            builder.addDOMMutator(doc -> {
                Element style = doc.createElement("style");
                Node node = doc.createTextNode(EXTRA_STYLE);
                style.appendChild(node);
                doc.getElementsByTagName("head").item(0).appendChild(style);
            });
            builder.run();
        }
    }`

**Comment by sunithapremkumar** [comment]
@danfickle would you be please able to quickly check my above code for displaying Chinese characters?

**Comment by sunithapremkumar** [comment]
I finally got it to work by the configuring the font family in right order in style (" 'Arial', 'noto', 'SimSun', 'sans-serif") and also using SimSun.ttf file for chinese as below as I couldn't find CJK ttf file (all I found was otf) which wasn't working.  I hope the font family order is correct please check.  

` private static final String EXTRA_STYLE =
            "* { font-family: 'Arial', 'noto', 'SimSun', 'sans-serif' !important; }\n" ;


    private static void convertHtmlToPDfusingOpenHtmlToPdf(String html) throws Exception{
        try (OutputStream os = new FileOutputStream("/output/main-report.pdf")) {
            PdfRendererBuilder builder = new PdfRendererBuilder();
            builder.useFastMode();
            builder.withHtmlContent(html, ".");
            builder.toStream(os);
            builder.useFont(new FontSupplier("/fonts/NotoSans-Regular.ttf"), "noto");
            builder.useFont(new FontSupplier("/fonts/SimSun.ttf"), "SimSun");
            builder.addDOMMutator(doc -> {
                Element style = doc.createElement("style");
                Node node = doc.createTextNode(EXTRA_STYLE);
                style.appendChild(node);
                doc.getElementsByTagName("head").item(0).appendChild(style);
            });

            builder.run();
        }`

**Comment by danfickle** [comment]
As for the first column being no wrap, I could only get everything to fit using landscape:

````java
    private static final String EXTRA_STYLE = 
        "* { font-family: 'noto', 'cjk' !important; }\n" +
        "table { table-layout: auto !important; }\n" +
        "td { white-space: normal !important;  width: auto !important; }\n" +
        "col { width: auto !important; }\n" +
        "td:first-child { white-space: nowrap !important; }\n" +
        "body td[colspan] { white-space: normal !important; }\n" +
        "body style { display: none !important; }\n" +
        "@page { size: A4 landscape; }\n";
````

**Comment by danfickle** [comment]
You can get [noto cjk here](https://github.com/danfickle/openhtmltopdf/issues/46#issuecomment-257242797).

**Comment by rkhlin** [comment]
Hi there,

I was using the Flying Saucer library before and it all went ok until I upgrade my Spring framework to v5 so that I have to look for a newer library to generate PDF with Chinese in it.

I had read quite a few posts and tried many ways, but still seeing ###. Not sure what is going on.
Please help.

My css file:
@CHARSET "UTF-8";
@font-face {
    font-family: "Arial Unicode MS";
    src: url('arialuni.ttf');
}
* {
  font-family: "Arial Unicode MS",sans-serif;
}

HTML file:
<!DOCTYPE html>
<html>
<head>
	<title>Testing</title>

</head>
<link rel="stylesheet" type="text/css" href="test.css" media="all" />
<body>
	<h1>每星期三<br/>
		Every Wednesday</h1>
</body>
</html>

Java:
	public static void main(String[] args) throws Exception {
        try (OutputStream os = new FileOutputStream("E:/Temp/out.pdf")) {
            PdfRendererBuilder builder = new PdfRendererBuilder();
//            builder.useFont(new File("E:/Temp/arialuni.ttf"),"Arial Unicode MS");
            builder.useFastMode();
            builder.withUri("file:///E:/Temp/index.html");
            builder.toStream(os);
            builder.run();
        }
    }

![out](https://user-images.githubusercontent.com/619188/132008032-64953643-e99b-4e1a-8d84-aadfbbe26912.jpg)
